README.md
# mediagrains

A python library for handling grain-based media in a python-native
style.

## Introduction

Provides constructor functions for various types of grains and classes
that nicely wrap those grains, as well as a full serialisation and
deserialisation library for GSF format. Please read the pydoc
documentation for more details.

Some useful tools for handling the Grain Sequence Format (GSF) file format
are also included - see [Tools](#tools).

## Installation

### Requirements

* A working Python 3.10+ installation
* BBC R&D's internal deb repository set up as a source for apt (if installing via apt-get)
* The tool [tox](https://tox.readthedocs.io/en/latest/) is needed to run the unittests, but not required to use the library.

### Steps

```bash
# Install from pip
$ pip install mediagrains

# Install via apt-get
$ apt-get install python3-mediagrains

# Install directly from source repo
$ git clone git@github.com:bbc/rd-apmm-python-lib-mediagrains.git
$ cd rd-apmm-python-lib-mediagrains
$ pip install -e .
```

## Usage

As an example of using this in your own code the following is a simple
program to load the contents of a GSF file and print the timestamp of
each grain.

```Python console
>>> from mediagrains.gsf import load
>>> f = open('examples/video.gsf', "rb")
>>> (head, segments) = load(f)
>>> print('\n'.join(str(grain.origin_timestamp) for grain in segments[1]))
1420102800:0
1420102800:20000000
1420102800:40000000
1420102800:60000000
1420102800:80000000
1420102800:100000000
1420102800:120000000
1420102800:140000000
1420102800:160000000
1420102800:180000000
```

Alternatively to create a new video grain in 10-bit planar YUV 4:2:0 and fill
it with colour-bars:

```Python console
>>> from mediagrains import VideoGrain
>>> from uuid import uuid1
>>> from mediagrains.cogenum import CogFrameFormat, CogFrameLayout
>>> src_id = uuid1()
>>> flow_id = uuid1()
>>> grain = VideoGrain(src_id, flow_id, cog_frame_format=CogFrameFormat.S16_422_10BIT, width=1920, height=1080)
>>> colours = [
...      (0x3FF, 0x000, 0x3FF),
...      (0x3FF, 0x3FF, 0x000),
...      (0x3FF, 0x000, 0x000),
...      (0x3FF, 0x3FF, 0x3FF),
...      (0x3FF, 0x200, 0x3FF),
...      (0x3FF, 0x3FF, 0x200) ]
>>> x_offset = [0, 0, 0]
>>> for colour in colours:
...     i = 0
...     for c in grain.components:
...             for x in range(0,c.width//len(colours)):
...                     for y in range(0,c.height):
...                             grain.data[c.offset + y*c.stride + (x_offset[i] + x)*2 + 0] = colour[i] & 0xFF
...                             grain.data[c.offset + y*c.stride + (x_offset[i] + x)*2 + 1] = colour[i] >> 8
...             x_offset[i] += c.width//len(colours)
...             i += 1
```

(a more natural interface for accessing data exists in the form of numpy arrays. See later.)

The object grain can then be freely used for whatever video processing
is desired, or it can be serialised into a GSF file as follows:

```Python console
>>> from mediagrains.gsf import dump
>>> f = open('dummyfile.gsf', 'wb')
>>> dump(f, [grain])
>>> f.close()
```
The encoding module also supports a "progressive" mode where an
encoder object is created and a dump started, then grains can be added
and will be written to the output file as they are added.

```Python console
>>> from uuid import uuid1
>>> from mediagrains import Grain
>>> from mediagrains.gsf import GSFEncoder
>>> src_id = uuid1()
>>> flow_id = uuid1()
>>> f = open('dummyfile.gsf', 'wb')
>>> enc = GSFEncoder(f)
>>> seg = enc.add_segment()  # This must be done before the call to start_dump
>>> enc.start_dump()  # This writes the file header and starts the export
>>> seg.add_grain(Grain(src_id, flow_id))  # Adds a grain and writes it to the file
>>> seg.add_grain(Grain(src_id, flow_id))  # Adds a grain and writes it to the file
>>> seg.add_grain(Grain(src_id, flow_id))  # Adds a grain and writes it to the file
>>> enc.end_dump()  # This ends the export and finishes off the file
>>> f.close()
```

If the underlying file is seekable then the end_dump call will upade all segment
metadata to list the correct grain count, otherwise the counts will be
left at -1.

In addition the library contains a relatively rich grain comparison
mechanism in the submodule `comparison`. An example of useage is as
follows:

```python
>>> from mediagrains.comparison import compare_grain
>>> from mediagrains.testsignalgenerator import LumaSteps
>>> from uuid import uuid1
>>> from mediatimestamp import Timestamp
>>> src_id = uuid1()
>>> flow_id = uuid1()
>>> ots = Timestamp()
>>> gen = LumaSteps(src_id, flow_id, 1920, 1080, origin_timestamp=ots)
>>> a = next(gen)
>>> b = next(gen)
>>> print(compare_grain(a, b))
❌   Grains do not match
  ✅   <a/b>.height == 1080
  ✅   Binary data <a/b>.data are equal
  ✅   <a/b>.format == <CogFrameFormat.U8_444: 8192>
  ✅   <a/b>.length == 6220800
  ❌   a.origin_timestamp - b.origin_timestamp == -0:40000000, not the expected 0:0
  ❌   a.sync_timestamp - b.sync_timestamp == -0:40000000, not the expected 0:0
  ✅   <a/b>.flow_id == UUID('7ff37130-d904-11e8-9fa5-5065f34ed007')
  ✅   <a/b>.grain_type == 'video'
  ✅   Lists match
    ✅   len(<a/b>.timelabels) == 0
  ✅   <a/b>.layout == <CogFrameLayout.FULL_FRAME: 0>
  ✅   <a/b>.width == 1920
  ✅   <a/b>.source_id == UUID('7bf845f6-d904-11e8-9fa5-5065f34ed007')
  ✅   <a/b>.rate == Fraction(25, 1)
  ✅   a.creation_timestamp - b.creation_timestamp == 0:0 as expected
  ✅   <a/b>.duration == Fraction(1, 25)
```

This output gives a relatively detailed breakdown of the differences
between two grains, both as a printed string (as seen above) and also
in a data-centric fashion as a tree structure which can be
interrogated in code.

### Numpy arrays

An additional feature is provided in the form of numpy array access to the data in a grain. As such the above example of creating colourbars can be done more easily:

```Python console
>>> from mediagrains.numpy import VideoGrain
>>> from uuid import uuid1
>>> from mediagrains.cogenums import CogFrameFormat, CogFrameLayout
>>> src_id = uuid1()
>>> flow_id = uuid1()
>>> grain = VideoGrain(src_id, flow_id, cog_frame_format=CogFrameFormat.S16_422_10BIT, width=1920, height=1080)
>>> colours = [
...      (0x3FF, 0x000, 0x3FF),
...      (0x3FF, 0x3FF, 0x000),
...      (0x3FF, 0x000, 0x000),
...      (0x3FF, 0x3FF, 0x3FF),
...      (0x3FF, 0x200, 0x3FF),
...      (0x3FF, 0x3FF, 0x200) ]
>>> for c in range(0, 3):
...     for x in range(0, grain.components[c].width):
...         for y in range(0, grain.components[c].height):
...             grain.component_data[c][x, y] = colours[x*len(colours)//grain.components[c].width][c]
```

## Documentation

The API is well documented in the docstrings of the module mediagrains, to view:

```bash
pydoc mediagrains
```

## Tools
Some tools are installed with the library to make working with the Grain Sequence Format (GSF) file format easier.

* `wrap_video_in_gsf` - Provides a means to read raw video essence and generate a GSF file.
* `wrap_audio_in_gsf` - As above, but for audio.
* `extract_from_gsf` - Read a GSF file and dump out the raw essence within.
* `gsf_probe` - Read metadata about the segments in a GSF file.

For example, to generate a GSF file containing a test pattern from `ffmpeg`, dump the metadata and then play it out
again:
```bash
ffmpeg -f lavfi -i testsrc=duration=20:size=1920x1080:rate=25 -pix_fmt yuv422p10le -c:v rawvideo -f rawvideo - | \
wrap_video_in_gsf - output.gsf --size 1920x1080 --format S16_422_10BIT --rate 25
gsf_probe output.gsf
extract_gsf_essence output.gsf - | ffplay -f rawvideo -pixel_format yuv422p10 -video_size 1920x1080 -framerate 25 pipe:0
```

To do the same with a sine wave:
```bash
ffmpeg -f lavfi -i "sine=frequency=1000:duration=5" -f s16le -ac 2 - | wrap_audio_in_gsf - output_audio.gsf --sample-rate 44100
gsf_probe output_audio.gsf
extract_gsf_essence output_audio.gsf - | ffplay -f s16le -ac 2 -ar 44100 pipe:0
```

## Development
### Testing

To run the unittests for this package in a virtual environment follow these steps:

```bash
$ git clone git@github.com:bbc/rd-apmm-python-lib-mediagrains.git
$ cd rd-apmm-python-lib-mediagrains
$ make test
```
### Packaging

Debian and RPM packages can be built using:

```bash
# Debian packaging
$ make deb

# RPM packageing
$ make rpm
```

### Continuous Integration

This repository includes a Jenkinsfile which makes use of custom steps defined in a BBC internal
library for use on our own Jenkins instances. As such it will not be immediately useable outside
of a BBC environment, but may still serve as inspiration and an example of how to implement CI
for this package.

## Versioning

We use [Semantic Versioning](https://semver.org/) for this repository

## Changelog

See [CHANGELOG.md](CHANGELOG.md)

## Contributing

All contributions are welcome, before submitting you must read and sign a copy of the [Individual Contributor License Agreement](ICLA.md)

Please ensure you have run the test suite before submitting a Pull Request, and include a version bump in line with our [Versioning](#versioning) policy.

## Authors

* James Weaver
* Philip deNier (philip.denier@bbc.co.uk)
* Sam Mesterton-Gibbons (sam.mesterton-gibbons@bbc.co.uk)
* Alex Rawcliffe
* James Sandford (james.sandford@bbc.co.uk)

## License

See [LICENSE.md](LICENSE.md)


CHANGELOG.md
# Mediagrains Library Changelog

## 3.0.0
- Support dropped for python <3.10 (3.8 should work but is not explicitely supported)
- `AUDIOGRAIN` and `VIDEOGRAIN` `data` field is now explicitely optional

## 2.16.0
- Add support for ADTS AAC (`--format AAC`) to the `wrap_audio_in_gsf` tool.

## 2.15.1
- BugFix: fix seek and tell in OpenAsyncStreamWrapper to be relative to the correct current position.

## 2.15.0
- Add support for H.264 (`--format H264`) to the `wrap_video_in_gsf` tool.

## 2.14.1
- Fix unbounded timerange in legacy test signal generators.

## 2.14.0
- Add a numpy AudioGrain.
  - All PCM audio formats other than the 24-bit ones can be modified.
- Add audio support to the `compute_psnr` and `compare_grain` functions.

## 2.13.0
- Deprecated testsignalgenerator
- Added patterngenerators

## 2.12.1
- Fix typing for the time value passed to the `normalise_time` method.

## 2.12.0
- Add `media_rate` property to GRAIN that returns video frame rate or audio sample rate.

## 2.11.0
- Added ParseGrainType as a type to describe a function that constructs grains

## 2.10.0
- Loosen restrictions on what can be used as a timestamp in grain construction, allowing the use of any object which supports the mediatimestamp protocol.

## 2.9.3
- Handle EOFError in `wrap_in_gsf` command line tools.

## 2.9.2
- Allow stdin as a valid input file for the tools

## 2.9.1
- Fix a bug where the `--samples_per_grain` option in `wrap_audio_in_gsf` wasn't honoured.

## 2.9.0
- Added CogAudioFormat.UNKNOWN which was erroneously missing from enum
- Add a flag for compairing grain iterators without retaining all grains in memory.

## 2.8.5
- Unpin mediatimestamp and mediajson

## 2.8.4
- Pin back versions of mediatimestamp and mediajson

## 2.8.3
- Bugfix: Numpy VideoGrain length should be the bytes length, not the array length.

## 2.8.2
- Removed code that used asyncio in synchronous calls to decode gsf and replaced it with a purely synchronous version.
  It turned out that since asyncio is fundementally incompatible with some third party libraries (eg. gevent) we need
  a code path that doesn't make any use of it.

## 2.8.1
- Added a bypass so that when wrapping a BytesIO in an Async wrapper unnecessary and potentially costly threads aren't spawned

## 2.8.0
- Switched to using asynctest for testing asynchronous code
- Made asynchronous IO wrappers that wrap synchronous IO do so using an executor thread pool
- Added support for automatically wrapping synchronous files in asynchronous wrappers to use them in gsf encoder
- Added bypass to wrap BytesIO in a lighter weight wrapper to prevent timing irregularities

## 2.7.2
- Bugfix: Restore behaviour whereby grains can be added to a segment object during an active progressive encode

## 2.7.1
- Bugfix: Restore behaviour whereby `gsf.GSFEncoder.dump` calls `gsf.GSFEncoder.start_dump` and `gsf.GSFEncoder.end_dump`
  (this matters if subclasses have overridden these methods)

## 2.7.0
- Dropped all support for Python2.7
- Moved python3.6 specific submodules in tree
- Added `GrainWrapper` class to wrap raw essence in Grains.
- Added `wrap_video_in_gsf` and `wrap_audio_in_gsf` tools to generate GSF files from raw essence.
- Added `extract_from_gsf` and `gsf_probe` tools to extract essence and metadata from GSF files.
- Added MyPy as a dependency
- Deprecated old asyncio code from v2.6
- Added Asynchronous GSFEncoding using the standard Encoder in a context-manager type workflow.
- Added Asynchronous GSFDecoding using the standard Decoder in a context-manager type workflow.

## 2.6.0
- Added support for async methods to gsf decoder in python 3.6+
- Added `Grain.origin_timerange` method.
- Added `Grain.normalise_time` method.
- Added `Colourbars` test signal generator
- Added `MovingBarOverlay` for test signal generators
- Added `mediagrains.numpy` sublibrary for handling video grains as numpy arrays, in python 3.6+
- Added `PSNR` option to grain compare.
- Support for converting between all uncompressed video grain formats added to `mediagrains.numpy`
- This is the last release that will support python 2.7 (apart from bugfixes)

## 2.5.3
- BUGFIX: IOBytes doesn't quite fulfil bytes-like contracts, but can be converted to something that does

## 2.5.2
- Remove dependency on enum34 for versions of python >= 3.4

## 2.5.1
- Fix restoring bytestream position when lazy loading bytes.

## 2.5.0
- Added ability to filter GSFDecoder output based on local ids

## 2.4.0
- Added ability for GSFDecoder to return grains with lazy loading
  bytes data

## 2.3.4
- Switch to using immutable timestamps throughout

## 2.3.3
- Allow build to run on a wider selection of agents.

## 2.3.2
- Added Jenkins trigger to rebuild master every day.

## 2.3.1
- Added upload of pydoc docs to Jenkinsfile

## 2.3.0
- Added Silence to test signal generation
- Added sample_rate option for audio test signal generators
- Added Tone test signal generator which can generate a vaiety of frequencies

## 2.2.0
- Added 1K Tone to test signal generation

## 2.1.0
- Added support for comparison of grains against other grains with a
  variety of options and detailed output when there are differences.

## 2.0.1
- Accept TimeOffsets as grain timestamps, to work around oddities in JSON
  parsing behaviour.

## 2.0.0
- GSFEncoder does not buffer grains once they have been written
- GSFEncoderSegment does not provide access to buffered grains

## 1.1.1
- Updated README.md

## 1.1.0
- GSFDecoder.grains() with skip_data=True positions stream pointer after data before yielding

## 1.0.0
- Initial Release version

## 0.4.0
- Add new `final_origin_timestamp` method to grains to make some logic
   simpler.
- Changed dependencies for timestamps to `mediatimestamp`.

## 0.3.1

mediagrains/Grain.py
#!/usr/bin/python
#
# Copyright 2018 British Broadcasting Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import print_function
from uuid import UUID
from nmoscommon.timestamp import Timestamp
from collections import Sequence, MutableSequence, Mapping
from fractions import Fraction
from cogframe import CogFrameFormat, CogFrameLayout

__all__ = ["Grain", "VideoGrain"]


class GRAIN(Sequence):
    """\
    A class representing a generic media grain.

    Any grain can be freely cast to a tuple:

      (meta, data)

    where meta is a dictionary containing the grain metadata, and data is a python buffer object representing the payload (or None for an empty grain).
    """
    def __init__(self, meta, data):
        self.meta = meta
        self.data = data
        self._factory = "Grain"

    def __len__(self):
        return 2

    def __getitem__(self, index):
        if index == 0:
            return self.meta
        elif index == 1:
            return self.data
        else:
            raise IndexError("tuple index out of range")

    def __repr__(self):
        if self.data is None:
            return "{}({!r})".format(self._factory, self.meta)
        else:
            return "{}({!r},{!r})".format(self._factory, self.meta, self.data)

    @property
    def grain_type(self):
        return self.meta['grain']['grain_type']

    @grain_type.setter
    def grain_type(self, value):
        self.meta['grain']['grain_type'] = value

    @property
    def source_id(self):
        return UUID(self.meta['grain']['source_id'])

    @source_id.setter
    def source_id(self, value):
        self.meta['grain']['source_id'] = str(value)

    @property
    def flow_id(self):
        return UUID(self.meta['grain']['flow_id'])

    @flow_id.setter
    def flow_id(self, value):
        self.meta['grain']['flow_id'] = str(value)

    @property
    def origin_timestamp(self):
        return Timestamp.from_tai_sec_nsec(self.meta['grain']['origin_timestamp'])

    @origin_timestamp.setter
    def origin_timestamp(self, value):
        if isinstance(value, Timestamp):
            value = value.to_tai_sec_nsec()
        self.meta['grain']['origin_timestamp'] = value

    @property
    def sync_timestamp(self):
        return Timestamp.from_tai_sec_nsec(self.meta['grain']['sync_timestamp'])

    @sync_timestamp.setter
    def sync_timestamp(self, value):
        if isinstance(value, Timestamp):
            value = value.to_tai_sec_nsec()
        self.meta['grain']['sync_timestamp'] = value

    @property
    def creation_timestamp(self):
        return Timestamp.from_tai_sec_nsec(self.meta['grain']['creation_timestamp'])

    @creation_timestamp.setter
    def creation_timestamp(self, value):
        if isinstance(value, Timestamp):
            value = value.to_tai_sec_nsec()
        self.meta['grain']['creation_timestamp'] = value

    @property
    def rate(self):
        return Fraction(self.meta['grain']['rate']['numerator'],
                        self.meta['grain']['rate']['denominator'])

    @rate.setter
    def rate(self, value):
        value = Fraction(value)
        self.meta['grain']['rate'] = {
            'numerator': value.numerator,
            'denominator': value.denominator
            }

    @property
    def duration(self):
        return Fraction(self.meta['grain']['duration']['numerator'],
                        self.meta['grain']['duration']['denominator'])

    @duration.setter
    def duration(self, value):
        value = Fraction(value)
        self.meta['grain']['duration'] = {
            'numerator': value.numerator,
            'denominator': value.denominator
            }

    @property
    def timelabels(self):
        if 'timelabels' in self.meta['grain']:
            return self.meta['grain']['timelabels']
        else:
            return []


class VIDEOGRAIN(GRAIN):

    class COMPONENT(Mapping):
        def __init__(self, meta):
            self.meta = meta

        def __getitem__(self, key):
            return self.meta[key]

        def __setitem__(self, key, value):
            self.meta[key] = value

        def __delitem__(self, key):
            del self.meta[key]

        def __iter__(self):
            return self.meta.__iter__()

        def __len__(self):
            return self.meta.__len__()

        @property
        def stride(self):
            return self.meta['stride']

        @stride.setter
        def stride(self, value):
            self.meta['stride'] = value

        @property
        def offset(self):
            return self.meta['offset']

        @offset.setter
        def offset(self, value):
            self.meta['offset'] = value

        @property
        def width(self):
            return self.meta['width']

        @width.setter
        def width(self, value):
            self.meta['width'] = value

        @property
        def height(self):
            return self.meta['height']

        @height.setter
        def height(self, value):
            self.meta['height'] = value

        @property
        def length(self):
            return self.meta['length']

        @length.setter
        def length(self, value):
            self.meta['length'] = value

    class COMPONENT_LIST(MutableSequence):
        def __init__(self, parent):
            self.parent = parent

        def __getitem__(self,key):
            return VIDEOGRAIN.COMPONENT(self.parent.meta['grain']['cog_frame']['components'][key])

        def __setitem__(self,key,value):
            self.parent.meta['grain']['cog_frame']['components'][key] = VIDEOGRAIN.COMPONENT(value)

        def __delitem__(self,key):
            del self.parent.meta['grain']['cog_frame']['components'][key]

        def insert(self, key, value):
            self.parent.meta['grain']['cog_frame']['components'].insert(key, VIDEOGRAIN.COMPONENT(value))

        def __len__(self):
            return len(self.parent.meta['grain']['cog_frame']['components'])

    def __init__(self, meta, data):
        super(VIDEOGRAIN, self).__init__(meta, data)
        self._factory = "VideoGrain"
        self.meta['grain']['grain_type'] = 'video'
        if 'cog_frame' not in self.meta['grain']:
            self.meta['grain']['cog_frame'] = {
                'format': CogFrameFormat.UNKNOWN,
                'width': 0,
                'height': 0,
                'layout': CogFrameLayout.UNKNOWN,
                'extension': 0,
                'components': []
            }
        if not isinstance(self.meta['grain']['cog_frame']['format'], CogFrameFormat):
            self.meta['grain']['cog_frame']['format'] = CogFrameFormat(self.meta['grain']['cog_frame']['format'])
        if not isinstance(self.meta['grain']['cog_frame']['layout'], CogFrameLayout):
            self.meta['grain']['cog_frame']['layout'] = CogFrameLayout(self.meta['grain']['cog_frame']['layout'])
        self.components = VIDEOGRAIN.COMPONENT_LIST(self)

    @property
    def format(self):
        return self.meta['grain']['cog_frame']['format']

    @format.setter
    def format(self, value):
        self.meta['grain']['cog_frame']['format'] = CogFrameFormat(value)

    @property
    def width(self):
        return self.meta['grain']['cog_frame']['width']

    @width.setter
    def width(self, value):
        self.meta['grain']['cog_frame']['width'] = value

    @property
    def height(self):
        return self.meta['grain']['cog_frame']['height']

    @height.setter
    def height(self, value):
        self.meta['grain']['cog_frame']['height'] = value

    @property
    def layout(self):
        return self.meta['grain']['cog_frame']['layout']

    @layout.setter
    def layout(self, value):
        self.meta['grain']['cog_frame']['layout'] = CogFrameLayout(value)

    @property
    def extension(self):
        return self.meta['grain']['cog_frame']['extension']

    @extension.setter
    def extension(self, value):
        self.meta['grain']['cog_frame']['extension'] = value

    @property
    def source_aspect_ratio(self):
        if 'source_aspect_ratio' in self.meta['grain']['cog_frame']:
            return Fraction(self.meta['grain']['cog_frame']['source_aspect_ratio']['numerator'],
                            self.meta['grain']['cog_frame']['source_aspect_ratio']['denominator'])
        else:
            return None

    @source_aspect_ratio.setter
    def source_aspect_ratio(self, value):
        value = Fraction(value)
        self.meta['grain']['cog_frame']['source_aspect_ratio'] = { 'numerator': value.numerator,
                                                          'denominator': value.denominator}

    @property
    def pixel_aspect_ratio(self):
        if 'pixel_aspect_ratio' in self.meta['grain']['cog_frame']:
            return Fraction(self.meta['grain']['cog_frame']['pixel_aspect_ratio']['numerator'],
                            self.meta['grain']['cog_frame']['pixel_aspect_ratio']['denominator'])
        else:
            return None

    @pixel_aspect_ratio.setter
    def pixel_aspect_ratio(self, value):
        value = Fraction(value)
        self.meta['grain']['cog_frame']['pixel_aspect_ratio'] = { 'numerator': value.numerator,
                                                         'denominator': value.denominator}


def size_for_format(fmt, w, h):
    if ((fmt>>8)&0x1) == 0x00:  # Cog frame is not packed
        h_shift = (fmt&0x01)
        v_shift = ((fmt>>1)&0x01)
        depth = (fmt&0xc)
        if depth == 0:
            bpv = 1
        elif depth == 4:
            bpv = 2
        else:
            bpv = 4

        return (w*h + 2*((w*h) >> (h_shift + v_shift)))*bpv
    else:
        if fmt in (CogFrameFormat.YUYV,
                   CogFrameFormat.UYVY,
                   CogFrameFormat.AYUV):
            return w*h*2
        elif fmt in (CogFrameFormat.RGBx,
                     CogFrameFormat.RGBA,
                     CogFrameFormat.xRGB,
                     CogFrameFormat.ARGB,
                     CogFrameFormat.BGRx,
                     CogFrameFormat.BGRA,
                     CogFrameFormat.xBGR,
                     CogFrameFormat.ABGR):
            return w*h*4
        elif fmt == CogFrameFormat.RGB:
            return w*h*3
        elif fmt == CogFrameFormat.v210:
            return h*(((w + 47) // 48) * 128)
        elif fmt == CogFrameFormat.v216:
            return w*h*4
        else:
            return 0

def components_for_format(fmt, w, h):
    components = []
    if ((fmt>>8)&0x1) == 0x00:  # Cog frame is not packed
        h_shift = (fmt&0x01)
        v_shift = ((fmt>>1)&0x01)
        depth = (fmt&0xc)
        if depth == 0:
            bpv = 1
        elif depth == 4:
            bpv = 2
        else:
            bpv = 4

        offset = 0
        components.append({
            'stride': w*bpv,
            'offset': offset,
            'width': w,
            'height': h,
            'length': w*h*bpv
        })
        offset += w*h*bpv

        components.append({
            'stride': (w >> h_shift)*bpv,
            'offset': offset,
            'width': w >> h_shift,
            'height': h >> v_shift,
            'length': ((w*h) >> (h_shift + v_shift))*bpv
        })
        offset += ((w*h) >> (h_shift + v_shift))*bpv

        components.append({
            'stride': (w >> h_shift)*bpv,
            'offset': offset,
            'width': w >> h_shift,
            'height': h >> v_shift,
            'length': ((w*h) >> (h_shift + v_shift))*bpv
        })
        offset += ((w*h) >> (h_shift + v_shift))*bpv

    else:
        if fmt in (CogFrameFormat.YUYV,
                   CogFrameFormat.UYVY,
                   CogFrameFormat.AYUV):
            components.append({
                'stride': w*2,
                'offset': 0,
                'width': w,
                'height': h,
                'length': h*w*2
            })
        elif fmt in (CogFrameFormat.RGBx,
                     CogFrameFormat.RGBA,
                     CogFrameFormat.xRGB,
                     CogFrameFormat.ARGB,
                     CogFrameFormat.BGRx,
                     CogFrameFormat.BGRA,
                     CogFrameFormat.xBGR,
                     CogFrameFormat.ABGR):
            components.append({
                'stride': w*4,
                'offset': 0,
                'width': w,
                'height': h,
                'length': h*w*4
            })
        elif fmt == CogFrameFormat.RGB:
            components.append({
                'stride': w*3,
                'offset': 0,
                'width': w,
                'height': h,
                'length': h*w*3
            })
        elif fmt == CogFrameFormat.v210:
            components.append({
                'stride': (((w + 47) // 48) * 128),
                'offset': 0,
                'width': w,
                'height': h,
                'length': h*(((w + 47) // 48) * 128)
            })
        elif fmt == CogFrameFormat.v216:
            components.append({
                'stride': w*4,
                'offset': 0,
                'width': w,
                'height': h,
                'length': h*w*4
            })
    return components


def VideoGrain(src_id_or_meta, flow_id_or_data=None, origin_timestamp=None,
               sync_timestamp=None, rate=Fraction(25,1), duration=Fraction(1,25),
               cog_frame_format=CogFrameFormat.UNKNOWN, width=1920,
               height=1080, cog_frame_layout=CogFrameLayout.UNKNOWN,
               flow_id=None, data=None):
    meta = None
    data = None
    src_id = None
    flow_id = None

    if isinstance(src_id_or_meta, dict):
        meta = src_id_or_meta
        if data is None:
            data = flow_id_or_data
    else:
        src_id = src_id_or_meta
        if flow_id is None:
            flow_id = flow_id_or_data

    if meta is None:
        if src_id is None or flow_id is None:
            raise AttributeError("Must include either metadata, or src_id, and flow_id")

        cts = Timestamp.get_time()
        if origin_timestamp is None:
            origin_timestamp = cts
        if sync_timestamp is None:
            sync_timestamp = origin_timestamp
        meta = {
            "@_ns": "urn:x-ipstudio:ns:0.1",
            "grain": {
                "grain_type": "video",
                "source_id": str(src_id),
                "flow_id": str(flow_id),
                "origin_timestamp": str(origin_timestamp),
                "sync_timestamp": str(sync_timestamp),
                "creation_timestamp": str(cts),
                "rate": {
                    "numerator": Fraction(rate).numerator,
                    "denominator": Fraction(rate).denominator,
                    },
                "duration": {
                    "numerator": Fraction(duration).numerator,
                    "denominator": Fraction(duration).denominator,
                    },
                "cog_frame": {
                    "format": cog_frame_format,
                    "width": width,
                    "height": height,
                    "layout": cog_frame_layout,
                    "extension": 0,
                    "components": []
                }
            },
        }

    if data is None:
        size = size_for_format(cog_frame_format, width, height)
        data = bytearray(size)

    if "cog_frame" in meta['grain'] and ("components" not in meta['grain']['cog_frame'] or len(meta['grain']['cog_frame']['components']) == 0):
        meta['grain']['cog_frame']['components'] = components_for_format(cog_frame_format, width, height)

    return VIDEOGRAIN(meta, data)

def Grain(src_id_or_meta=None, flow_id_or_data=None, origin_timestamp=None,
          sync_timestamp=None, rate=Fraction(25,1), duration=Fraction(1,25),
          flow_id=None, data=None, src_id=None, meta=None):
    """\
    Several possible ways to construct a grain:

      Grain(src_id, flow_id, origin_timestamp=current_time,
            sync_timestamp=origin_timestamp)

    creates a new empty grain in the specified source id (uuid.UUID or string)
    and flow_id

      Grain(meta, [ data ])

    creates a new grain with the specified data and optional buffer object
    for payload
"""

    if meta is None:
        if isinstance(src_id_or_meta, dict):
            meta = src_id_or_meta
            if data is None:
                data = flow_id_or_data
        else:
            src_id = src_id_or_meta
            if flow_id is None:
                flow_id = flow_id_or_data

    if meta is None:
        cts = Timestamp.get_time()
        ots = origin_timestamp
        sts = sync_timestamp

        if ots is None:
            ots = cts
        if sts is None:
            sts = ots

        if src_id is None or flow_id is None:
            raise AttributeError("Must specify at least meta or src_id and flow_id")

        if isinstance(src_id, UUID):
            src_id = str(src_id)
        if isinstance(flow_id, UUID):
            flow_id = str(flow_id)

        if not isinstance(src_id, basestring) or not isinstance(flow_id, basestring):
            raise AttributeError("Invalid types for src_id and flow_id")

        meta = {
            "@_ns": "urn:x-ipstudio:ns:0.1",
            "grain": {
                "grain_type": "empty",
                "source_id": src_id,
                "flow_id": flow_id,
                "origin_timestamp": str(ots),
                "sync_timestamp": str(sts),
                "creation_timestamp": str(cts),
                "rate": {
                    "numerator": 0,
                    "denominator": 1,
                    },
                "duration": {
                    "numerator": 0,
                    "denominator": 1,
                    },
                }
            }
        data = None

    if meta['grain']['grain_type'] == 'video':
        return VideoGrain(meta, data)
    else:
        return GRAIN(meta, data)


if __name__ == "__main__":  # pragma: no cover
    from uuid import uuid1, uuid5

    src_id = uuid1()
    flow_id = uuid5(src_id, "flow_id:test_flow")

    grain1 = Grain(src_id, flow_id)
    grain2 = Grain(grain1.meta)
    print(grain1)
    print(grain2)
